// @fleekhq/browser-rpc/dist/PortRPC.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const extensionizer_1 = __importDefault(require("extensionizer"));
const RPC_1 = __importDefault(require("./RPC"));
class PortRPC extends RPC_1.default {
    constructor(config) {
        super(config);
        this.port = undefined;
        this.receiveMessage = this.receiveMessage.bind(this);
    }
    sendMessage(message) {
        if (this.port) {
            this.port.postMessage(message);
        }
    }
    receiveMessage(message) {
        super.onMessage(message);
    }
    start() {
        this.port = extensionizer_1.default.runtime.connect({ name: this.name });
        if (this.port) {
            this.port.onMessage.addListener(this.receiveMessage);
        }
    }
    stop() {
        super.stop();
        if (this.port) {
            this.port.onMessage.removeListener(this.receiveMessage);
        }
    }
}
;
exports.default = PortRPC;
