// @fleekhq/browser-rpc/dist/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("joi"));
const schema = joi_1.default.object({
    target: joi_1.default.string().required(),
    data: joi_1.default.object({
        name: joi_1.default.string().required(),
        data: joi_1.default.object({
            id: joi_1.default.alternatives().try(joi_1.default.number(), joi_1.default.string()).required(),
            jsonrpc: joi_1.default.string().valid('2.0').required(),
            result: joi_1.default.any(),
            method: joi_1.default.string(),
            params: joi_1.default.array().items(joi_1.default.any()),
            error: joi_1.default.object({
                code: joi_1.default.number().required(),
                message: joi_1.default.string().required(),
                data: joi_1.default.any(),
            }),
        }).required(),
    }).required(),
});
exports.default = schema;
